### PR TITLE
Create new Substreams sinks documentation

### PR DESCRIPTION
**Pull request by seanmooretechwriter**
Create new page and content for Substreams sinks including links to resources, code snippets, explanations, and guidance.

### CODE REVIEW

**Review by maoueh — CHANGES_REQUESTED** [review]
Pushing my first message of the review, see our channel for longer feedback

**Comment by seanmooretechwriter** [comment]
I'm beginning this now. I'll start by replacing the current content with the outline you sent.

**Comment by seanmooretechwriter** [comment]
New outline proposed in Slack.


*Introduction*
In this paragraph, I would like that when the reader reads this, he gets out with the feeling that he as full freedom to "sink" Substreams to anything they want. A sink could even be to send message to Slack directly. So, database is not broad enough, sink is more than that. Let's have the user dream here about the possibilities.

Needs to be re-written to show all the possibilities people how people could have sinks. To database, to file format, custom processing, Slack message, name it.

The StreamingFasst methodology of development referes to databases as “sinks.”

I don't like the wording of The StreamingFasst methodology here, feels non-casual.


*Basics*
Actual content re-written to not be database centric


*General Requirements*
General considerations that sink accepts a particular module’s output, each sink defines it’s own set of requirements around that. Some accepts a single type (Postgres, Mongo), some others accepts multiple output types (sink-files).


*Existing Sinks*
Call to actions to other articles about existing Sink articles
Community Sink


*Build your own Sink*
Describe what users can do if a sink for there “preferred” tool does not exist
Derived way to achieve same results (for example sink to files than ingest in BigQuery)
Quick overview of how one can built one from scratch (use substreams-sink library and use a simple one as a template)

**Comment by maoueh** [comment]
Yep go ahead, reach out to me for any details you would like to have to write the content

**Comment by seanmooretechwriter** [comment]
Here is the first paragraph. Please let me know if this is acceptable.
-----------------------

The data captured from a blockchain with Substreams can be routed to multiple types of sinks. A sink is an end container to house the data acquired. Examples are a database, Slack channel, or flat file storage. Sinks have a wide range of types and Substreams data can be routed anywhere a developer can imagine.

StreamingFast provides a few examples, libraries, and tools to assist Substreams developers with routing blockchain data to sinks.

**Comment by seanmooretechwriter** [comment]
Here is database-less content for the second section as requested.

Data captured and processed by Substreams can be stored in many different ways through sinks. A Substreams developer’s imagination is really the only limitation. Immediate and typical storage types could be a database or flat files however Substreams data can be piped into other desired locations required by a new or existing application or architecture.

An important design aspect of Substreams is the deciiosn to rely on Google Protocol Buffers, or protobufs, for data packaging and transmission. Protobufs provide a data-centric, technology stack and languages agnostic approach to working with data that is passed from one application to another. The application-agnostic capabilities of protobufs give developers the opportunity to package and route data captured by Substreams to other sources, including sinks.

One of the critical steps involved is the creation of a protobuf that forms data to meet the requirements of a sink. The protobuf is populated with blockchain data captured in a Substreams module and then used as output. Existing sink solutions, such as PostgreSQL, provided by StreamingFast, demonstrate this functionality. It’s important to note that databases are merely one type of sink. 

The consuming application, or code, can read protobuf-based data being sent out of Substreams. Protobufs are flexible and the expectations of the consuming application can be matched closely with mindful data design. Substreams will send the data through a map module using a protobuf defined by the developer. The data is then consumed by another application that will route the data to the desired location, or sink.

An understanding of basic Substreams fundamentals is suggested before continuing. Learn more about modules basics in the Substreams documentation at the following link.

https://substreams.streamingfast.io/concept-and-fundamentals/modules

**Comment by seanmooretechwriter** [comment]
*Build your own Sink*
"Describe what users can do if a sink for there “preferred” tool does not exist"

Here is content for review for this piece. 
-------------------------

StreamingFast provides tools allowing developers to route blockchain data to a few different types of data storage sinks, or means of ingestion. The types of sinks with tools provided by StreamingFast aren’t the only options for Substreams developers. Existing applications, databases, and other tools can be fed by blockchain data captured and output by Substreams.

Developers can examine the StreamingFast sink tools to see examples of how protobuf-based data from Substreams can be used with other approaches. One example could be a database, such as Oracle, that doesn’t currently have tools in place. As mentioned, databases are only one example of where Substreams data can be routed. Developers should be able to review the PostgreSQL tool and its codebase to begin to understand how to construct their own data-sinking solution.

Reiterating from above, protobufs are designed by the developer. The protobufs are used to transfer data out of Substreams to the data sink. Protobufs aren’t tied to any particular technology stack or language, enabling developers to capture, further process, use and store the Substreams data in a myriad of different capacities.

Through careful design of the Substreams manifest, modules, and protobufs developers can craft their output data in many ways. One option, as seen in the PostgreSQL example is through a single output protobuf. The flexibility of Substreams design however allows for other strategies, including multiple protobufs and modules. Developers need to examine and account for the format and any requirements of the end target they want their data routed. The specifics of how data is ingested by the targeted sink will determine the design of the output from Substreams.
----------------------------------------------
 
"Derived way to achieve same results (for example sink to files than ingest in BigQuery)" -- I'll need further input on this if it's not been addressed in the content provided above. 

"Quick overview of how one can built one from scratch (use substreams-sink library and use a simple one as a template)" -- I'll begin working on this next.

**Comment by maoueh** [comment]
> Community Sink -- I'm not sure what this is. Can you provide a link or additional information, please?

There is none today, but here it would be maybe a list of some sink maintained outside of StreamingFast team. For now, let's put a call to actions, something in the vein of. "We value external contributions for Substreams Sink (bla bla). You have created a Sink, reach to us so we can add it here".

Something in those lines, better written.

**Comment by maoueh** [comment]
The rest seems pretty good, let's turn that into proper markdown files.

> Build your own Sink

Good starting point, would like to see links to existing implementation. We also created a library `substreams-sink` to build `sink` most easily.

Later, we could think about you creating an easy one (maybe invoking a bash script each time some data is received), that way we could give more detailed instructions about how creating a `sink`.

**Comment by seanmooretechwriter** [comment]
Ok, thank you for the input. This is helpful!

I'll work on these changes. Thank you.

**Comment by seanmooretechwriter** [comment]
I updated the markdown file, committed the changes, and pushed to this branch for review.

File: docs/developer-guide/substreams-sinks.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+An important design aspect of Substreams is the deciiosn to rely on Google Protocol Buffers, or protobufs, for data packaging and transmission. Protobufs provide a data-centric, technology stack and languages agnostic approach to working with data that is passed from one application to another. The application-agnostic capabilities of protobufs give developers the opportunity to package and route data captured by Substreams to other sources, including sinks.
+
+## **General Requirements**
+
+One of the critical steps involved is the creation of a protobuf that forms data to meet the requirements of a sink. The protobuf is populated with blockchain data captured in a Substreams module and then used as output. Existing sink solutions, such as PostgreSQL, provided by StreamingFast, demonstrate this functionality. It’s important to note that databases are merely one type of sink.
```
Comment (maoueh) [inline docs/developer-guide/substreams-sinks.md:17]:
```suggestion
The first step of having your Substreams ready to be consumed by a particular Sink involves the creation of a `map` module whose output type is a Protobuf accepted by it. This specific Protobuf is populated from your own Substreams Protobuf types, you can see this as a transformation of your types into a format suitable for Sink to ingest.

For example, database-likes Substreams Sinks such as PostgreSQL or MongoDB accepts a module's output of type [substreams.database.v1.DatabaseChanges](https://github.com/streamingfast/substreams-database-change/blob/develop/proto/database/v1/database.proto#L5).

It’s important to note that databases are merely one type of sink, the sink you are targeting determines what output type you should respect.
```

File: docs/developer-guide/substreams-sinks.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+The consuming application, or code, can read protobuf-based data being sent out of Substreams. Protobufs are flexible and the expectations of the consuming application can be matched closely with mindful data design. Substreams will send the data through a map module using a protobuf defined by the developer. The data is then consumed by another application that will route the data to the desired location, or sink.
```
Comment (maoueh) [inline docs/developer-guide/substreams-sinks.md:19]:
```suggestion
The sink reads the specific protobuf-based data being sent out of Substreams and performs the processing for it. Every sink performs differently regarding the data received, most of them performing some kind of storage. 

The configuration of this storage layer and the requirements of it is your responsibility. Each Substreams Sink should document the specific infrastructure required for running the `sink`. You should read the documentation about the particular sink you are using to understand correctly its behavior and requirements.
```

**Review by maoueh — CHANGES_REQUESTED** [review]
Ok, leaving my review comments. Those are suggestions, don't hesitate to re-word to read better, you have a very nice prose so my humble writing could benefits from it.

File: docs/developer-guide/substreams-sinks.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Substreams Sinks
+
+## **Introduction**
+
+The data captured from a blockchain with Substreams can be routed to multiple types of sinks. A sink is an end container to house the data acquired. Examples are a database, Slack channel, or flat file storage. Sinks have a wide range of types and Substreams data can be routed anywhere a developer can imagine.
```
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:5]: Could we change this to something that explains a sink as simply the "final destination" for the data retrieved from the blockchain via Substreams? Is that technically accurate?

File: docs/developer-guide/substreams-sinks.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+StreamingFast provides a few examples, libraries, and tools to assist Substreams developers with routing blockchain data to sinks.
+
+## **Basics**
+
+Data captured and processed by Substreams can be stored in many different ways through sinks. A Substreams developer’s imagination is really the only limitation. Immediate and typical storage types could be a database or flat files however Substreams data can be piped into other desired locations required by a new or existing application or architecture.
+
+An important design aspect of Substreams is the deciiosn to rely on Google Protocol Buffers, or protobufs, for data packaging and transmission. Protobufs provide a data-centric, technology stack and languages agnostic approach to working with data that is passed from one application to another. The application-agnostic capabilities of protobufs give developers the opportunity to package and route data captured by Substreams to other sources, including sinks.
+
```
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:14]: Addressed in file. Will be available for review in the forthcoming PR.

File: docs/developer-guide/substreams-sinks.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+One of the critical steps involved is the creation of a protobuf that forms data to meet the requirements of a sink. The protobuf is populated with blockchain data captured in a Substreams module and then used as output. Existing sink solutions, such as PostgreSQL, provided by StreamingFast, demonstrate this functionality. It’s important to note that databases are merely one type of sink.
+
+The consuming application, or code, can read protobuf-based data being sent out of Substreams. Protobufs are flexible and the expectations of the consuming application can be matched closely with mindful data design. Substreams will send the data through a map module using a protobuf defined by the developer. The data is then consumed by another application that will route the data to the desired location, or sink.
+
+An understanding of basic Substreams fundamentals is suggested before continuing. Learn more about modules basics in the Substreams documentation at the following link.
+
+https://substreams.streamingfast.io/concept-and-fundamentals/modules
+
+## **Existing & Commnuity Sinks**
+
```
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:32]: Addressed in file. Will be available for review in the forthcoming PR.

File: docs/developer-guide/substreams-sinks.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+Through careful design of the Substreams manifest, modules, and protobufs developers can craft their output data in many ways. One option, as seen in the PostgreSQL example is through a single output protobuf. The flexibility of Substreams design however allows for other strategies, including multiple protobufs and modules. Developers need to examine and account for the format and any requirements of the end target they want their data routed. The specifics of how data is ingested by the targeted sink will determine the design of the output from Substreams.
+
+The substreams-eth-block-meta example demonstrates sinks in action. Check out the source code in the project’s official GitHub repository.
+
+https://github.com/streamingfast/substreams-eth-block-meta
+
+StreamingFast provides several tools to assist Substreams developers interested in persisting data to databases; each can be found in its official GitHub repository.
```
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:49]: Addressed in file. Will be available for review in the forthcoming PR.

File: docs/developer-guide/substreams-sinks.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+## **Build a Sink**
+
+StreamingFast provides tools allowing developers to route blockchain data to a few different types of data storage sinks, or means of ingestion. The types of sinks with tools provided by StreamingFast aren’t the only options for Substreams developers. Existing applications, databases, and other tools can be fed by blockchain data captured and output by Substreams.
+
+Developers can examine the StreamingFast sink tools to see examples of how protobuf-based data from Substreams can be used with other approaches. One example could be a database, such as Oracle, that doesn’t currently have tools in place. As mentioned, databases are only one example of where Substreams data can be routed. Developers should be able to review the PostgreSQL tool and its codebase to begin to understand how to construct their own data-sinking solution.
+
+Reiterating from above, protobufs are designed by the developer. The protobufs are used to transfer data out of Substreams to the data sink. Protobufs aren’t tied to any particular technology stack or language, enabling developers to capture, further process, use and store the Substreams data in a myriad of different capacities.
```
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:41]: Addressed in file. Will be available for review in the forthcoming PR.

File: docs/developer-guide/substreams-sinks.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:5]: All other changes have been addressed. Please let me know if the language change from "storage" to " final destination" works for defining sinks in the context. I'll make the update and push the code for the final review.

Thanks!!!!

[the remark this answers is not in the document]
Comment (maoueh) [reply docs/developer-guide/substreams-sinks.md:5]: I'm good with `final destination`, my goal here is really to show that other forms of processing can exists that are not storage. Altering system could be fed for example from a substreams sink which does not involve storage at all.

[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md:5]: Ok, that makes sense. Also, Slack wouldn't really be a standard or typical data store.

Got it. Thank you. I'll make the updates.

[the remark this answers is not in the document]
Comment (seanmooretechwriter) [reply docs/developer-guide/substreams-sinks.md]: Ok, this has been addressed, updated, committed, and pushed.

**Pull request merged**